File: week2/omerSefaYilmaz/generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
    const array = [];
  
    const buildParenthesis = (openCount, closeCount, str, n, arr) => {
      if (str.length === n * 2) {
        arr.push(str);
        return;
      }
      if (openCount)
        buildParenthesis(openCount - 1, closeCount, str + "(", n, arr);
      if (closeCount > openCount)
        buildParenthesis(openCount, closeCount - 1, str + ")", n, arr);
    };
    buildParenthesis(n, n, "", n, array);
    return array;
  };
  

File: week2/esadmsk_w2/generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        def backtrack(combination, open_count, close_count):
            if len(combination) == 2 * n:
                result.append(combination)
                return
            if open_count < n:
                backtrack(combination + '(', open_count + 1, close_count)
            if close_count < open_count:
                backtrack(combination + ')', open_count, close_count + 1)

        result = []
        backtrack('', 0, 0)
        return result

